✅ 662 jest eventBus helpers

paduc opened the pull request:
Introduces 3 helpers pour testing `EventBus` behavior

- `expect(eventBus).toHavePublished(Event)`
- `expect(eventBus).toHavePublishedTimes(n)`
- `expect(eventBus).toHavePublishedWithPayload(Event, payload)` (with __partial__ match on payload)

JulienPavon reviewed (changes requested):
Si le test échoue le résultat n'est pas très parlant. D'ailleurs il semble y avoir une erreur :

```
● authN.handleUserCreated › should give rights to all existing projects with this email

    TypeError: pretty_format_1.format is not a function

      41 |           const difference = diffLinesUnified2(
      42 |             // serialize with indentation to display lines
    > 43 |             format(receivedPayload).split('\n'),
         |             ^
      44 |             format(expectedPayload).split('\n'),
      45 |             // serialize without indentation to compare lines
      46 |             format(receivedPayload, { indent: 0 }).split('\n'),

      at message (src/__tests__/matchers/toHavePublishedWithPayload.ts:43:13)
```

Du coup je me demande si utiliser expect.objectContaining dans un matcher toHaveBeenNthCalledWith ne suffirait-il pas ?

Par exemple :
```
expect(eventBus.publish).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        type: 'UserProjectsLinkedByContactEmail',
        payload: {
          userId: 'userId',
          projectIds: [projectWithSameEmailId],
        },
      })
    )
```
Ici on ne vérifierait donc que ce que nous souhaitons sans les champs problématiques comme la date de l'event

paduc commented:
> Si le test échoue le résultat n'est pas très parlant. D'ailleurs il semble y avoir une erreur :

Est-ce que tu as fait un `npm install` ?

Tu veux dire le résultat du test qui utiliserait `toHavePublishWithPayload` ou bien le test de `toHavePublishedWithPayload` lui-même ?

> 
> ```
> ● authN.handleUserCreated › should give rights to all existing projects with this email
> 
>     TypeError: pretty_format_1.format is not a function
> 
>       41 |           const difference = diffLinesUnified2(
>       42 |             // serialize with indentation to display lines
>     > 43 |             format(receivedPayload).split('\n'),
>          |             ^
>       44 |             format(expectedPayload).split('\n'),
>       45 |             // serialize without indentation to compare lines
>       46 |             format(receivedPayload, { indent: 0 }).split('\n'),
> 
>       at message (src/__tests__/matchers/toHavePublishedWithPayload.ts:43:13)
> ```
> 
> Du coup je me demande si utiliser expect.objectContaining dans un matcher toHaveBeenNthCalledWith ne suffirait-il pas ?
> 
> Par exemple :
> 
> ```
> expect(eventBus.publish).toHaveBeenNthCalledWith(
>       1,
>       expect.objectContaining({
>         type: 'UserProjectsLinkedByContactEmail',
>         payload: {
>           userId: 'userId',
>           projectIds: [projectWithSameEmailId],
>         },
>       })
>     )
> ```
> 
> Ici on ne vérifierait donc que ce que nous souhaitons sans les champs problématiques comme la date de l'event

Oui on pourrait faire ça en effet. L'idée de ce helper est d'être plus explicite / proche de l'archi.

JulienPavon commented:
> > Si le test échoue le résultat n'est pas très parlant. D'ailleurs il semble y avoir une erreur :
> 
> Est-ce que tu as fait un `npm install` ?
> 
> Tu veux dire le résultat du test qui utiliserait `toHavePublishWithPayload` ou bien le test de `toHavePublishedWithPayload` lui-même ?

Effectivement le npm install était manquant. C'est ok maintneant. et c'était pour le test qui utilisait toHavePublishWithPayload.
